### PR TITLE
fix: initial value in FAST select control not used when using Angular two-way binding

### LINKED ISSUE #5139: fix: initial value in FAST select control not used when using Angular two-way binding
# 🐛 Bug Report

When using a FAST select control in Angular, a two-way value binding does not set the initial value of the control. I'm using [ngDefaultControl](https://angular.io/api/forms/DefaultValueAccessor), as suggested in the [FAST Angular integration docs](https://www.fast.design/docs/integrations/angular). 

## 💻 Repro or Code Sample

https://stackblitz.com/edit/angular-ivy-7kjj4p?file=src/app/app.component.html

## 🤔 Expected Behavior

When the select's value is two-way bound to a value in the component using `[(ngModel)]`, the initial value of the control should be the same as the initial value of the property in the component.

## 😯 Current Behavior

The select control does not use the value in the bound property for its initial value; instead it uses the first value in the listbox. 

## 💁 Possible Solution

It looks like the select's `value` property is being set before the `slottedOptions` property, so it treats the value as invalid. I don't see an obvious fix though.

## 🔦 Context

Trying the use the FAST select control in Angular with nice two-way binding behavior and form integration.

## 🌍 Your Environment

* OS & Device: Windows 10
* Browser Microsoft Edge
* Version 92.0.902.78

### PR DESCRIPTION
**Pull request by brianehenry**
# Pull Request

## 📖 Description

The FAST select control fails to accept its initial value when it is set right after the select is connected. This change handles that case by marking the `option` element corresponding to the given value with the `selected` attribute. 

### 🎫 Issues

Fixes #5139 

## 📑 Test Plan

Added one unit test that reflects the desired behavior.

### General

- [x] I have included a change request file using `$ yarn change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

### CODE REVIEW

**Review by radium-v — CHANGES_REQUESTED** [review]
Hi @brianehenry, thanks for investigating this issue. I downloaded and ran the angular demo from #5139 to make sure my review is accurate.

The approach works for setting the default selected option when `ngModel` is used, but there are two problems:
* It sets the `selected` attribute on the matching `fast-option`, which can cause collisions when both a default `value` exists and an initial child option already has the `selected` attribute.
* It runs the check every time `value` is updated.

Since we only need it to check this when the component is connected, we can override the `setDefaultSelectedOption()` method for `Select` in the same way that `Combobox` does it.

This override function is a modified version from `Listbox`, which allows the options list to fallback to the filtered list of child elements, and adds a `value` check for each option:

```ts
    protected setDefaultSelectedOption(): void {
        if (this.$fastController.isConnected) {
            const options: ListboxOption[] =
                this.options ??
                Array.from(this.children).filter(Listbox.slottedOptionFilter);

            const selectedIndex = options?.findIndex(
                el =>
                    el.getAttribute("selected") !== null ||
                    el.selected ||
                    el.value === this.value
            );

            if (selectedIndex !== -1) {
                this.selectedIndex = selectedIndex;
                return;
            }

            this.selectedIndex = 0;
        }
    }
```

Then to make sure the `formResetCallback` functions as expected, we can change the method to call the super method:

```ts
public formResetCallback = (): void => {
    this.setProxyOptions();
    super.setDefaultSelectedOption();
    this.value = this.firstSelectedOption.value;
};
```

**Comment by brianehenry** [comment]
Hi @radium-v, thanks for the detailed feedback. 

I tried out your suggested changes, but they don't seem to fix the issue that I'm seeing (or, at least, the test case I wrote fails with those changes). The issue with your suggestion is that it relies on the select's private `_value` field being set by the `value` setter. But the `value` setter does not set `_value` if it does not find a matching `option` element with the same `value` string, which is the case when the `options` property has not yet been initialized. So, when `setDefaultSelectedOption()` is called, the select's `value` property is still unset.

I updated my PR to address the first issue you had with my changes - the select will now clear any previously set `selected` attribute on the option elements. I don't think the second issue you raised is a showstopper, as the `value` setter won't call any of the code I added after the `options` property is initialized.

**Comment by EisenbergEffect** [comment]
@radium-v Can you take a look at this one again. Brian has made some changes since your last feedback.

**Comment by brianehenry** [comment]
@radium-v Provided some additional feedback [on discord](https://discord.com/channels/449251231706251264/449251231706251266/931692250679873547). I applied the suggested changes to my branch.

**Review by EisenbergEffect — APPROVED** [review]
Just need a bit of code commenting for the on out-of-the-ordinary situation noted below.

File: packages/web-components/fast-foundation/src/select/select.ts [hunk 1]
```diff
@@ -305,6 +306,25 @@ export class Select extends FormAssociatedSelect {
         this.updateValue();
     }
 
+    protected setDefaultSelectedOption(): void {
+        const options: ListboxOption[] =
+            this.options ?? Array.from(this.children).filter(Listbox.slottedOptionFilter);
+
+        const selectedIndex = options?.findIndex(
+            el =>
+                el.getAttribute("selected") !== null ||
```
Comment (radium-v) [inline packages/web-components/fast-foundation/src/select/select.ts:315]: My bad, there's a better check for this:

```suggestion
                el.hasAttribute("selected") ||
```

**Review by radium-v — APPROVED** [review]

File: packages/web-components/fast-foundation/src/select/select.ts [hunk 2]
```diff
@@ -229,7 +230,7 @@ export class Select extends FormAssociatedSelect {
      */
     public formResetCallback(): void {
         this.setProxyOptions();
-        this.setDefaultSelectedOption();
+        super.setDefaultSelectedOption();
```
[the remark this answers is not in the document]
Comment (brianehenry) [reply packages/web-components/fast-foundation/src/select/select.ts:235]: Added a comment with an explanation. For what it's worth, I find it strange to call a `super` method from anywhere besides its override, but I don't see an obvious alternative here.

**Comment by brianehenry** [comment]
@chrisdholt @EisenbergEffect This PR is ready to be merged.

**Pull request merged**